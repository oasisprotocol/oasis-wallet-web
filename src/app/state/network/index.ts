import { PayloadAction } from '@reduxjs/toolkit'
import { createSlice } from 'utils/@reduxjs/toolkit'
import { NetworkState, NetworkType } from './types'

export const initialState: NetworkState = {
  ticker: '',
  chainContext: '',
  selectedNetwork: process.env.REACT_APP_LOCALNET ? 'local' : 'mainnet',
  epoch: 0,
  minimumStakingAmount: 0,
}

export const networkSlice = createSlice({
  name: 'network',
  initialState,
  reducers: {
    initializeNetwork() {},
    selectNetwork(state, action: PayloadAction<NetworkType>) {
      state.chainContext = initialState.chainContext
      state.epoch = initialState.epoch
    },
    initialNetworkSelected(state, action: PayloadAction<NetworkState>) {
      Object.assign(state, action.payload)
    },
    networkSelected(state, action: PayloadAction<NetworkState>) {
      Object.assign(state, action.payload, {
        chainContext: initialState.chainContext,
        epoch: initialState.epoch,
      })
    },
    setChainContext(state, action: PayloadAction<string>) {
      state.chainContext = action.payload
    },
    setEpoch(state, action: PayloadAction<number>) {
      state.epoch = action.payload
    },
    getEpoch(state) {},
  },
})

export const { actions: networkActions } = networkSlice
