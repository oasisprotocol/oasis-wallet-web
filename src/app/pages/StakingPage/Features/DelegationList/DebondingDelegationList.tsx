import { useEffect } from 'react'
import { selectDebondingDelegations } from 'app/state/staking/selectors'
import { Box } from 'grommet/es6/components/Box'
import { useDispatch, useSelector } from 'react-redux'

import { DelegationList } from '.'
import { StakeSubnavigation } from '../../../AccountPage/Features/StakeSubnavigation'
import { networkActions } from '../../../../../app/state/network'
import { selectEpoch } from '../../../../../app/state/network/selectors'

export const DebondingDelegationList = () => {
  const dispatch = useDispatch()
  const delegations = useSelector(selectDebondingDelegations)
  const currentEpoch = useSelector(selectEpoch)

  useEffect(() => {
    return () => {
      dispatch(networkActions.getEpoch())
    }
  }, [dispatch])

  return (
    <>
      <StakeSubnavigation />
      <Box as="section" data-testid="debonding-delegations">
        <Box pad="medium" background="background-front">
          <DelegationList currentEpoch={currentEpoch} type="debonding" delegations={delegations ?? []} />
        </Box>
      </Box>
    </>
  )
}
