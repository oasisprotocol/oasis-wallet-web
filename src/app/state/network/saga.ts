import * as oasis from '@oasisprotocol/client'
import { persistActions } from 'app/state/persist'
import { selectSkipUnlockingOnInit } from 'app/state/persist/selectors'
import { config } from 'config'
import { RECEIVE_INIT_STATE } from 'redux-state-sync'
import { call, put, select, takeLatest } from 'typed-redux-saga'
import { backend, backendApi } from 'vendors/backend'

import { networkActions } from '.'
import { SyncedRootState } from '../persist/types'
import { selectChainContext, selectEpoch, selectSelectedNetwork } from './selectors'
import { NetworkType } from './types'
import { WalletError, WalletErrors } from 'types/errors'

/**
 * Return a nic client for the specified network,
 * or by default, for the currently selected network
 */
export function* getOasisNic(network?: NetworkType) {
  const selectedNetwork = network ? network : yield* select(selectSelectedNetwork)
  const url = config[selectedNetwork].grpc

  const nic = new oasis.client.NodeInternal(url)
  return nic
}

/**
 * Return the explorer APIs for the currently selected network
 */
export function* getExplorerAPIs() {
  const selectedNetwork = yield* select(selectSelectedNetwork)
  const url = config[selectedNetwork][backend()].explorer
  return backendApi(url)
}

export function* getChainContext() {
  const chainContext = yield* select(selectChainContext)
  if (chainContext) {
    return chainContext
  }

  try {
    const selectedNetwork = yield* select(selectSelectedNetwork)
    const nic = yield* call(getOasisNic, selectedNetwork)
    const fetchedChainContext = yield* call([nic, nic.consensusGetChainContext])
    yield* put(networkActions.setChainContext(fetchedChainContext))
    return fetchedChainContext
  } catch (error) {
    throw new WalletError(WalletErrors.UnknownGrpcError, 'Could not fetch data')
  }
}

export function* getEpoch() {
  const epoch = yield* select(selectEpoch)
  if (epoch) {
    return epoch
  }

  try {
    const selectedNetwork = yield* select(selectSelectedNetwork)
    const nic = yield* call(getOasisNic, selectedNetwork)
    const fetchedEpoch = yield* call([nic, nic.beaconGetEpoch], oasis.consensus.HEIGHT_LATEST)
    const fetchedEpochNumber = Number(fetchedEpoch)
    yield* put(networkActions.setEpoch(fetchedEpochNumber))
    return fetchedEpochNumber
  } catch (error) {
    throw new WalletError(WalletErrors.UnknownGrpcError, 'Could not fetch data')
  }
}

export function* selectNetwork({
  network,
  isInitializing,
}: {
  network: NetworkType
  isInitializing: boolean
}) {
  const networkState = {
    ticker: config[network].ticker,
    selectedNetwork: network,
    minimumStakingAmount: config[network].min_delegation,
  }

  if (isInitializing) {
    yield* put(networkActions.initialNetworkSelected(networkState))
  } else {
    yield* put(networkActions.networkSelected(networkState))
  }
}

export function* networkSaga() {
  yield* takeLatest(
    [
      networkActions.initializeNetwork,
      networkActions.selectNetwork,
      persistActions.setUnlockedRootState,
      persistActions.resetRootState,
      RECEIVE_INIT_STATE,
    ],
    function* (action) {
      const currentNetworkType = yield* select(selectSelectedNetwork)

      if (networkActions.initializeNetwork.match(action)) {
        yield* call(selectNetwork, { network: currentNetworkType, isInitializing: true })
      }
      if (networkActions.selectNetwork.match(action)) {
        yield* call(selectNetwork, { network: action.payload, isInitializing: false })
      }
      if (persistActions.setUnlockedRootState.match(action)) {
        const rootState = action.payload.persistedRootState
        yield* call(selectNetwork, { network: rootState.network.selectedNetwork, isInitializing: false })
      }
      if (persistActions.resetRootState.match(action)) {
        yield* call(selectNetwork, { network: currentNetworkType, isInitializing: false })
      }
      if (action.type === RECEIVE_INIT_STATE) {
        const rootState = (action as any).payload as SyncedRootState
        yield* call(selectNetwork, { network: rootState.network.selectedNetwork, isInitializing: false })
      }
    },
  )
  yield* takeLatest(persistActions.resetRootState, function* () {
    const skipUnlockOnInit = yield* select(selectSkipUnlockingOnInit)
    if (skipUnlockOnInit) {
      yield* put(persistActions.skipUnlocking())
    }
  })

  yield* takeLatest(networkActions.getEpoch, getEpoch)

  yield* put(networkActions.initializeNetwork())
}
