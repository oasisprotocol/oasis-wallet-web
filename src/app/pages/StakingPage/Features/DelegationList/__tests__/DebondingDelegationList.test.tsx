import userEvent from '@testing-library/user-event'
import { act, screen, waitFor } from '@testing-library/react'
import * as React from 'react'
import { render } from '@testing-library/react'

import { Provider } from 'react-redux'
import { DebondingDelegationList } from '../DebondingDelegationList'
import { configureAppStore } from 'store/configureStore'
import { stakingActions } from 'app/state/staking'
import { NetworkState } from '../../../../../../app/state/network/types'
import { ThemeProvider } from '../../../../../../styles/theme/ThemeProvider'
import { MemoryRouter } from 'react-router-dom'

const renderComponent = (store: any) =>
  render(
    <Provider store={store}>
      <ThemeProvider>
        <MemoryRouter>
          <DebondingDelegationList />
        </MemoryRouter>
      </ThemeProvider>
    </Provider>,
  )

describe('<DebondingDelegationList  />', () => {
  let store: ReturnType<typeof configureAppStore>

  beforeEach(() => {
    store = configureAppStore({
      network: {
        ticker: 'TEST',
        epoch: 4,
      } as NetworkState,
    })
  })

  it('should match snapshot', () => {
    renderComponent(store)
    act(() => {
      store.dispatch(
        stakingActions.updateDelegations({
          delegations: [],
          debondingDelegations: [
            {
              epoch: 100,
              amount: 100n.toString(),
              shares: 100n.toString(),
              validatorAddress: 'test-validator',
              validator: {
                current_rate: 0.1,
                address: 'test-validator',
                rank: 1,
                status: 'active',
                name: 'test-validator',
                nodeAddress: 'oasis1qq7pgk9v8l3hu2aenjtflezy5vajc2cz3y4d96rj',
                escrow: 1000n.toString(),
              },
            },
          ],
        }),
      )
    })

    expect(screen.getByTestId('debonding-delegations')).toMatchSnapshot()
  })

  it('should expand and display the delegation on click', async () => {
    renderComponent(store)
    act(() => {
      store.dispatch(
        stakingActions.updateDelegations({
          delegations: [],
          debondingDelegations: [
            {
              epoch: 100,
              amount: 100n.toString(),
              shares: 100n.toString(),
              validatorAddress: 'oasis1qqv25adrld8jjquzxzg769689lgf9jxvwgjs8tha',
              validator: {
                address: 'oasis1qqv25adrld8jjquzxzg769689lgf9jxvwgjs8tha',
                current_rate: 0,
                rank: 1,
                status: 'active',
                name: 'test-validator1',
                nodeAddress: 'oasis1qq7pgk9v8l3hu2aenjtflezy5vajc2cz3y4d96rj',
                escrow: 1000n.toString(),
              },
            },
            {
              amount: 50n.toString(),
              shares: 50n.toString(),
              validatorAddress: 'oasis1qq2vzcvxn0js5unsch5me2xz4kr43vcasv0d5eq4',
              epoch: 100,
            },
          ],
        }),
      )
    })

    let row = screen.getByText(/test-validator1/)
    expect(row).toBeVisible()
    await userEvent.click(row)

    const details = await screen.findByTestId('validator-item')
    row = screen.getAllByText(/test-validator1/)[0]

    expect(details).toBeVisible()
    await userEvent.click(row)
    await waitFor(() => expect(details).not.toBeVisible())
  })
})
