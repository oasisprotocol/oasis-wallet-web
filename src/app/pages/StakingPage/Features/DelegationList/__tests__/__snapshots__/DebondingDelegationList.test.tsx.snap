// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DebondingDelegationList  /> should match snapshot 1`] = `
.c11 {
  display: inline-block;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  width: 20px;
  height: 20px;
  fill: currentColor;
  stroke: currentColor;
}

.c11 g {
  fill: inherit;
  stroke: inherit;
}

.c11 *:not([stroke])[fill='none'] {
  stroke-width: 0;
}

.c11 *[stroke*='#'],
.c11 *[STROKE*='#'] {
  stroke: inherit;
  fill: none;
}

.c11 *[fill-rule],
.c11 *[FILL-RULE],
.c11 *[fill*='#'],
.c11 *[FILL*='#'] {
  fill: inherit;
  stroke: none;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  background-color: #FFFFFF;
  color: #16213E;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding: 24px;
}

.c19 {
  font-size: 16px;
  line-height: 20px;
  color: #0500e2;
}

.c4 {
  position: relative;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100%;
  height: 100%;
  max-width: 100%;
  color: unset;
  background-color: unset;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  text-align: left;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  width: 100%;
  background-color: unset;
  min-height: 56px;
  border-bottom-width: 1px;
  border-bottom-color: #AAAAAAaa;
  border-bottom-style: solid;
}

.c7 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-sizing: border-box;
  line-height: normal;
  font-size: 16px;
  font-weight: 500;
  color: unset;
  padding-left: 16px;
  padding-right: 16px;
}

.c17 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-sizing: border-box;
  line-height: normal;
  padding-left: 16px;
  padding-right: 16px;
  word-break: break-word;
}

.c8 {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-flex-basis: 0;
  -ms-flex-preferred-size: 0;
  flex-basis: 0;
  max-width: 100%;
  min-width: 100px;
  min-width: 34px;
  max-width: 34px;
}

.c9 {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-flex-basis: 0;
  -ms-flex-preferred-size: 0;
  flex-basis: 0;
  max-width: 40ex;
  min-width: 15ex;
}

.c12 {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-flex-basis: 0;
  -ms-flex-preferred-size: 0;
  flex-basis: 0;
  max-width: 100%;
  min-width: 100px;
  min-width: 28ex;
  max-width: 28ex;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

.c13 {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-flex-basis: 0;
  -ms-flex-preferred-size: 0;
  flex-basis: 0;
  max-width: 100%;
  min-width: 100px;
}

.c18 {
  font-weight: 400;
}

.c16 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  width: 100%;
  box-sizing: border-box;
  font-size: 16px;
  color: unset;
  background-color: unset;
  min-height: 48px;
}

.c16:not(:last-of-type) {
  border-bottom-style: solid;
  border-bottom-width: 1px;
  border-bottom-color: #AAAAAAaa;
}

.c16:hover {
  color: unset;
  background-color: #88888833;
  -webkit-transition-duration: 0.15s;
  transition-duration: 0.15s;
  -webkit-transition-property: background-color;
  transition-property: background-color;
  border-bottom-color: unset;
  outline-style: solid;
  outline-width: 1px;
  outline-color: unset;
}

.c16:hover {
  cursor: pointer;
}

.c10 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 100%;
  line-height: 1;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c10:focus {
  outline: none;
  color: unset;
}

.c10:hover {
  color: unset;
}

.c10 span.__rdt_custom_sort_icon__ i,
.c10 span.__rdt_custom_sort_icon__ svg {
  opacity: 0;
  color: inherit;
  font-size: 18px !important;
  height: 18px !important;
  width: 18px !important;
  -webkit-backface-visibility: hidden;
  backface-visibility: hidden;
  -webkit-transform-style: preserve-3d;
  -ms-transform-style: preserve-3d;
  transform-style: preserve-3d;
  -webkit-transition-duration: 125ms;
  transition-duration: 125ms;
  -webkit-transition-property: -webkit-transform;
  -webkit-transition-property: transform;
  transition-property: transform;
}

.c10 span.__rdt_custom_sort_icon__.asc i,
.c10 span.__rdt_custom_sort_icon__.asc svg {
  -webkit-transform: rotate(180deg);
  -ms-transform: rotate(180deg);
  transform: rotate(180deg);
}

.c10:hover {
  cursor: pointer;
}

.c10:hover span,
.c10:hover span.__rdt_custom_sort_icon__ * {
  opacity: 1;
}

.c14 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 100%;
  line-height: 1;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  color: unset;
}

.c14:focus {
  outline: none;
}

.c14:hover:not(:focus) {
  color: unset;
}

.c14 span.__rdt_custom_sort_icon__ i,
.c14 span.__rdt_custom_sort_icon__ svg {
  opacity: 1;
  color: inherit;
  font-size: 18px !important;
  height: 18px !important;
  width: 18px !important;
  -webkit-backface-visibility: hidden;
  backface-visibility: hidden;
  -webkit-transform-style: preserve-3d;
  -ms-transform-style: preserve-3d;
  transform-style: preserve-3d;
  -webkit-transition-duration: 125ms;
  transition-duration: 125ms;
  -webkit-transition-property: -webkit-transform;
  -webkit-transition-property: transform;
  transition-property: transform;
}

.c14 span.__rdt_custom_sort_icon__.asc i,
.c14 span.__rdt_custom_sort_icon__.asc svg {
  -webkit-transform: rotate(180deg);
  -ms-transform: rotate(180deg);
  transform: rotate(180deg);
}

.c14:hover {
  cursor: pointer;
}

.c15 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c2 {
  position: relative;
  width: 100%;
  border-radius: inherit;
  overflow-x: auto;
  overflow-y: hidden;
  min-height: 0;
}

.c3 {
  position: relative;
  width: 100%;
  display: table;
}

@media only screen and (max-width:768px) {

}

@media only screen and (max-width:768px) {
  .c1 {
    padding: 12px;
  }
}

@media only screen and (max-width:768px) {

}

@media only screen and (max-width:768px) {

}

<section
  class="c0"
  data-testid="debonding-delegations"
>
  <div
    class="c1"
  >
    <div
      class="c2"
    >
      <div
        class="c3"
      >
        <div
          class="c4 rdt_Table"
          role="table"
        >
          <div
            class="c5 rdt_TableHead"
            role="rowgroup"
          >
            <div
              class="c6 rdt_TableHeadRow"
              role="row"
            >
              <div
                class="c7 c8 rdt_TableCol"
              />
              <div
                class="c7 c9 rdt_TableCol"
              >
                <div
                  class="c10 rdt_TableCol_Sortable"
                  id="column-name"
                  role="columnheader"
                  tabindex="0"
                >
                  <div>
                    validator.name
                  </div>
                  <span
                    class="asc __rdt_custom_sort_icon__"
                  >
                    <svg
                      aria-label="Down"
                      class="c11"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="m2 8.35 10.173 9.823L21.997 8"
                        fill="none"
                        stroke="#000"
                        stroke-width="2"
                      />
                    </svg>
                  </span>
                </div>
              </div>
              <div
                class="c7 c12 rdt_TableCol"
              >
                <div
                  class="c10 rdt_TableCol_Sortable"
                  id="column-amount"
                  role="columnheader"
                  tabindex="0"
                >
                  <span
                    class="asc __rdt_custom_sort_icon__"
                  >
                    <svg
                      aria-label="Down"
                      class="c11"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="m2 8.35 10.173 9.823L21.997 8"
                        fill="none"
                        stroke="#000"
                        stroke-width="2"
                      />
                    </svg>
                  </span>
                  <div>
                    delegations.reclaimedAmount
                  </div>
                </div>
              </div>
              <div
                class="c7 c13 rdt_TableCol"
              >
                <div
                  class="c14 rdt_TableCol_Sortable"
                  id="column-epoch"
                  role="columnheader"
                  tabindex="0"
                >
                  <div>
                    delegations.debondingTimeEnd
                  </div>
                  <span
                    class="asc __rdt_custom_sort_icon__"
                  >
                    <svg
                      aria-label="Down"
                      class="c11"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="m2 8.35 10.173 9.823L21.997 8"
                        fill="none"
                        stroke="#000"
                        stroke-width="2"
                      />
                    </svg>
                  </span>
                </div>
              </div>
            </div>
          </div>
          <div
            class="c15 rdt_TableBody"
            offset="250px"
            role="rowgroup"
          >
            <div
              class="c16 rdt_TableRow"
              id="row-test-validator+100"
              role="row"
            >
              <div
                class="c17 c8 c18 rdt_TableCell"
                data-tag="allowRowEvents"
                id="cell-icon-test-validator+100"
                role="gridcell"
              >
                <div />
              </div>
              <div
                class="c17 c9 c18 rdt_TableCell"
                data-tag="allowRowEvents"
                id="cell-name-test-validator+100"
                role="gridcell"
              >
                <div>
                  test-validator
                </div>
              </div>
              <div
                class="c17 c12 c18 rdt_TableCell"
                data-tag="allowRowEvents"
                id="cell-amount-test-validator+100"
                role="gridcell"
              >
                <div>
                  <span>
                    <div
                      class="c0"
                      style="display: inline-flex; white-space: nowrap; font-family: "Roboto mono", monospace; letter-spacing: 0;"
                    >
                      0.00
                    </div>
                    <span
                      class="c19"
                    >
                      <span
                        class="notranslate"
                        translate="no"
                      >
                         TEST
                      </span>
                    </span>
                  </span>
                </div>
              </div>
              <div
                class="c17 c13 c18 rdt_TableCell"
                data-tag="allowRowEvents"
                id="cell-debondingTimeEnd-test-validator+100"
                role="gridcell"
              >
                <div>
                  <span>
                    in 4 days
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</section>
`;
